consensus/istanbul/core/handler.go

```diff
@@ -162,6 +168,74 @@ func (c *core) sendEvent(ev interface{}) {
 	c.backend.EventMux().Post(ev)
 }
 
+// Returns true if this message is considered a future message.
+func IsFutureMsg(currentSequence, desiredRound *big.Int, currentState State, msgView *istanbul.View, msgCode uint64) bool {
+
+	// Future sequences indicate a future message.
+	if msgView.Sequence.Cmp(currentSequence) > 0 {
+		return true
+	}
+	// Future rounds inidcate a future message except for round changes
+	if msgView.Round.Cmp(desiredRound) > 0 && msgCode != istanbul.MsgRoundChange {
+		return true
+	}
+	// If we are waiting for a proposal or a proposed value then commit and
+	// prpepare messages are considered future.
+	if currentState.In(StateWaitingForNewRound, StateAcceptRequest) &&
+		(msgCode == istanbul.MsgCommit || msgCode == istanbul.MsgPrepare) {
+		return true
+	}
+	return false
+}
+
+// buildAlgorithmMsg builds an algorithm message from an istanbul message. It
+// takes the oracle because it needs to validate round change certificates and
+// generate an associated id. This is pretty clunky, ideally we could represent
+// the round change certificate in more detail in the algorithm message and
+// then the round change certificate verification logic could be moved to the
+// algorithm, but this will be a bit of work.
+func buildAlgorithmMsg(msg *istanbul.Message, o *RoundStateOracle) (*algorithm.Msg, error) {
```

Comment:
Not sure i see a lot of value on this. I see value in having a method that no matter the msg is able to give me the msg's round and sequence, and if there's another thing that is shared among messages. But besides that, it does not provid emuch